NUX: extend the NUX SiteType UI component to include a free form text option.

AnnaMag opened the pull request:
Change applicable to only Jetpack sites.
The change is temporary and is meant to serve the purpose of gathering user data.

#### Changes proposed in this Pull Request

* Add extra field to the form asking for site type during onboarding.

#### Testing instructions
- at `http://calypso.localhost:3000/jetpack/connect/site-type/:connected_site`  verify that the form has an extra field (as per mockup blow).
- chose Other option, input some text in the field and submit the form
- in /options.php of the site, verify the that `site_topic` option contains text `other- your_text`.

- at `http://calypso.localhost:3000/start/onboarding/site-type`  verify that the form has NO extra fields

<img width="521" alt="Screenshot 2019-04-01 at 12 18 52" src="https://user-images.githubusercontent.com/13561163/55320859-7ce98c00-5478-11e9-8dee-77042cfdb3c2.png">


<img width="390" alt="Screenshot 2019-03-26 at 15 15 20" src="https://user-images.githubusercontent.com/13561163/55004226-1f69c100-4fda-11e9-9430-456fa24cd949.png">

shaunandrews commented:
This feels very heavy. What about keeping the list of options simple, but if you select "Other" you go to a new step that lets you enter your site-type?

Or, if we really wanna keep it on one step, I'd imagine this should use a small, single-line input rather than a big textarea.

shaunandrews commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -10,6 +10,7 @@ import { localize } from 'i18n-calypso';
  * Internal dependencies
  */
 import Card from 'components/card';
+import CountedTextarea from 'components/forms/counted-textarea';
```
I don't think the character count is helpful in this use-case. Perhaps use a simple, single-line text input. Or consider moving this to a new step that would be shown after the customer selects "Other".

AnnaMag replied:
I would like to limit text length, encourage words and avoid prose (for easy text parsing). Else, the data ceases to be very useful.

AnnaMag replied:
what we did for the disconnect flow (no char limit was not a great move ):
<img width="705" alt="Screenshot 2019-03-26 at 17 17 00" src="https://user-images.githubusercontent.com/13561163/55014307-0584aa00-4feb-11e9-9ed8-44128743e367.png">

fditrapani commented:
> This feels very heavy. What about keeping the list of options simple, but if you select "Other" you go to a new step that lets you enter your site-type?

I agree the text area and character count feels heavy. Shaun, for some context @keoshi did explore an option that went to another step but that didn't feel quite right either. I always imagined doing something progressive in that when you pressed the other option, the field would be revealed.

AnnaMag commented:
@shaunandrews yeah, it does feel clunky, thus putting this here as a conversation starter. 
@jeffgolenski is working on the refinement atm.

I'd like to avoid messy data due to:
- too long text (we want keywords, not sentences)
- user confusion between expected info in the site type vs topic entries.
Thus, presenting it on the same screen.

Q: I was under impression that this extra field would apply to JP only (relevant code change pending conformation). @fditrapani could you comment on whether the wpcom sites will have the change live also?

fditrapani commented:
> Q: I was under impression that this extra field would apply to JP only (relevant code change pending conformation). @fditrapani could you comment on whether the wpcom sites will have the change live also?

I thought we were going to do both but I think it's fine if we don't since these options were created based on inputs collected from WP.com. You can go ahead and apply this to Jetpack only. Thanks for checking in.

jeffgolenski commented:
@keoshi @fditrapani @shaunandrews What if we were to break down the component a bit and keep the transitions right inside the "other" line item. We could implement a simple text input. 

I tried to do a fancy prototype with transitions in figma, but just ended up getting extremely frustrated. So just envision the sweet transitions shaun and filipe typically conjure, but limited to just inside the last item of this list.

![Frame 01](https://user-images.githubusercontent.com/214813/55023134-8a16fe80-4fd2-11e9-997a-b4406d0b2073.png)


![Frame 02](https://user-images.githubusercontent.com/214813/55023149-8edbb280-4fd2-11e9-9040-58a8ddaf4b3f.png)


![Frame 03](https://user-images.githubusercontent.com/214813/55023163-9307d000-4fd2-11e9-92d8-38234e93c59c.png)

shaunandrews commented:
I like the simplicity, but wonder how you'd submit the input — there's no submit button.

jeffgolenski commented:
@shaunandrews the right facing chevron would turn green once text is added to the input as a "next" indicator.

shaunandrews commented:
I think a button would be needed. The chevron is used to denote that a row is clickable. Having it become a button breaks the pattern and isn't obvious.

Here's a few ideas building on your concept.
--
Like you outlined, we'd include an input in the Other row, but not a chevron:
<img width="436" alt="image" src="https://user-images.githubusercontent.com/191598/55025721-e29cca80-4fd7-11e9-8cfb-a500604b697d.png">

Once you start typing in the input, we'd hide the "Other" label and show a "Continue" button:
<img width="447" alt="image" src="https://user-images.githubusercontent.com/191598/55025815-0eb84b80-4fd8-11e9-8d5f-81a64689dd66.png">

--

Another option would be to offer the other input as a separate part of the step-content:
<img width="883" alt="image" src="https://user-images.githubusercontent.com/191598/55025872-2a235680-4fd8-11e9-94ed-8a7884d57b94.png">

AnnaMag commented:
@shaunandrews the arrow works as a submit button (as in other cases). The patch is working btw :)

jeffgolenski commented:
@shaunandrews I think your second iteration is a great quick win. "Or type your own." What do you think about limiting the redundancy of the label / placeholder text?

shaunandrews commented:
> What do you think about limiting the redundancy of the label / placeholder text?

Sure, whatever ya'll think is best. This was a 30s mockup. :)

AnnaMag commented:
@shaunandrews how is the text submitted in the 2nd mockup: via enter or CTA appearing once users start typing?

keoshi commented:
Not going to comment on any specific design approach but I wanted to bring back something from the discussion over on https://github.com/Automattic/wp-calypso/issues/31560:

Are we trying to fit users into a mold or do we want them to tell us exactly what site type they're building/connecting?

I reckon we started with great defaults, but if we want to iterate on the existing list with real-world data that only our customers can provide us with, we shouldn't be worried about drawing too much attention to the input.

I also think that to some degree we don't need an input here, to be honest. We're already collecting data on the site topic/vertical step and I believe most customers won't understand the nuance between the two questions (heck, having two of them might even annoy them in the first place). In that sense, having them click “Other” and input their custom vertical on the second step would be a clearer path for them, and for us as we collect data and iterate on this experience.

shaunandrews commented:
> In that sense, having them click “Other” and input their custom vertical on the second step would be a clearer path for them, and for us as we collect data and iterate on this experience.

100% behind this. If we get gather the data we need without adding any UI at all, that seems like a great solution.

AnnaMag commented:
Thank you all for the input.
+100 for drawing attention to experiment/use the extra field. The main goal is to encourage the use of it. 

Atm, we have good pass through rates and having an extra input, if we make it optional (skip with no compulsory) + merge the Skip button, should not feel like a trap/forced UX. For data analysis, I'd like to see both free form inputs in action.  
Exemplary doubt: what if the site type input brings us clarity while the vertical one confuses users (aside from easier NLP)? We are second-guessing, while from looking at the data site vertical description could possibly be more specific. Q for a separate issue/PR: could we adapt the copy of "Tell us about your website" to encourage less description?

AnnaMag commented:
Updated view:
<img width="521" alt="Screenshot 2019-04-01 at 12 18 52" src="https://user-images.githubusercontent.com/13561163/55320835-6c391600-5478-11e9-8d4c-0398d6cd1273.png">

AnnaMag commented:
@shaunandrews I've included the design changes you proposed above in the form. I left the CTA button active at all times to keep the "skipping" option available. The idea is to make it optional and avoid trapping users into forced surveys.

Could you please re-test and possibly adjust if required (if you have bandwith) any faulty responsiveness? I'm not sure the effect is as-requested atm.

simison commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,24 +52,60 @@ class SiteTypeForm extends Component {
+				</Card>
+				<LoggedOutFormLinks className="site-type__links">
+					<LoggedOutFormLinkItem className="site-type__text">
+						Or type your own
```
This is missing `translate()`

AnnaMag replied:
ofc!

shaunandrews commented:
Working up some design tweaks for this.

shaunandrews commented:
Rebased and pushed a few style/markup tweaks. Here's how it's looking:

<img width="607" alt="image" src="https://user-images.githubusercontent.com/191598/55419476-d3230180-5542-11e9-8584-28a7c628e5a4.png">

--

The only thing left, I think, is that the button is clickable when the input is empty — but nothing happens. A few options:

1. Hide the button until you start typing
2. Show a disabled button, and then enable when you start typing
3. Show an error when submitting a blank input
4. Allow blank replies, and treat it like skipping the step

jeffgolenski commented:
@shaunandrews I'm loving this iteration. 

Would it make sense to match the design pattern of the site vertical / topic step and just have the button disabled until something is typed in the box? I'm worried people might see the green and automatically click "continue" without realizing they can click / tap on the list above. Just an assumption at this point, tho!

AnnaMag commented:
@jeffgolenski just a note (in case it adds more context) that the initial motivation behind the button being active is to offer an Other (no text) option, as suggested by Filipe above -> https://github.com/Automattic/wp-calypso/pull/31524#issuecomment-477047955. This would indicate reason without giving specifics, yet provide an alternative that indicates "none of the above".

AnnaMag commented:
@tyxla this PR is ready for another review when/if you have a chance.

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +52,88 @@ class SiteTypeForm extends Component {
-				key={ siteTypeProperties.id }
-				displayAsLink
-				data-e2e-title={ siteTypeProperties.slug }
-				tagName="button"
```
Question: We seem to be changing the `<button />` to `<div />` - could you please explain the rationale behind this change? If there's no reason for it, It looks like an accessibility regression.

AnnaMag replied:
Added back.

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +52,88 @@ class SiteTypeForm extends Component {
+
+		return (
+			<div className="site-type__other-option">
+				<p className="site-type__other-label">{ translate( 'Or type your own' ) }</p>
```
I think this should actually be a `<FormLabel />`, and upon click, to focus the field below.

AnnaMag replied:
I reckon it is static, explanatory text.

tyxla replied:
Let's ask a designer just in case :) @jeffgolenski what do you think?

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +52,88 @@ class SiteTypeForm extends Component {
-		return <Card className="site-type__wrapper">{ this.renderSegmentOptions() }</Card>;
+		const { isJetpack } = this.props;
+
+		if ( isJetpack === null ) {
```
This will break if this step eventually gets displayed on a route with a site slug for a non-Jetpack site. In that case, `isJetpack` will be `false`. So I suggest making this a simple bool check to go around this case.

```suggestion
		if ( ! isJetpack ) {
```

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -10,7 +10,11 @@ import { localize } from 'i18n-calypso';
  * Internal dependencies
  */
 import Card from 'components/card';
+import Button from 'components/forms/form-button';
```
Nit: would be great to keep imports alphabetized 😉

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -20,19 +24,23 @@ import './style.scss';
-	constructor( props ) {
-		super( props );
-		this.state = {
-			siteType: props.siteType,
```
Hm, the WP.com step still relies on this, so we shouldn't be removing it:

https://github.com/Automattic/wp-calypso/blob/ad2e4bdb62ead2751cc40e5c14a6e30b3195b6d2/client/signup/steps/site-type/index.jsx#L53

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +52,88 @@ class SiteTypeForm extends Component {
+		this.handleSubmit( 'other—' + this.state.otherValue );
+	};
+
+	setOtherReasonFocus = focus => () => {
```
This doesn't seem to be used. Should we remove, or use?

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +52,88 @@ class SiteTypeForm extends Component {
 
 export default connect(
-	null,
+	state => {
```
Looks like we could simplify:

```
state => {
	const siteId = getSelectedSiteId( state );
	const isJetpack = isJetpackSite( state, siteId );
	return {
		isJetpack,
	};
},
```

to:

```
state => ( {
	isJetpack: isJetpackSite( state, getSelectedSiteId( state ) ),
},
```

tyxla commented on client/signup/steps/site-type/style.scss:
```diff
@@ -78,3 +77,52 @@
 	}
 }
+
+.site-type__other-option {
```
I'd love to remove some of this CSS by using `<FormTextInputWithAction />` instead of building our own version.

AnnaMag replied:
I appreciate the styling affinity suggestions. This code is temporary (will be removed shortly), so in order to support its main goal (data collection) and make use of it asap, I'll merge it as is. To reiterate: we will remove this code in less than a month.

tyxla replied:
Well, while I'm fine to trust your judgement, I disagree. 

Just to explain: my problem is that these duplications remain forever - they don't seem to get addressed, and then as we move to other projects, we forget about them. FWIW that was the case for the user type step, but it still stays duplicated to this day. I just disagree we should keep setting bad precedents for shipping unfinished code and leaving it as-is. Ideally, we should reuse the rich component base we already have, and not introduce obsolete, duplicate components and/or styles in the first place, which I'm trying to do here.

Maybe we should ask for a second opinion - someone from @Automattic/luna or @Automattic/team-calypso - I'm fine to agree with the opinion of the majority.

AnnaMag replied:
We record issues to be address. Recent events have been pretty impacting re the priorities.

> I just disagree we should keep setting bad precedents for shipping unfinished code and leaving it as-is.

Again, I do not consider it fair to make generalized statements. It seems like you are referring to a larger problem and other situations. 
I disagree with discarding deadlines in the name of perfection—we need to test our assumptions efficiently.

tyxla replied:
As stated in my [latest review](https://github.com/Automattic/wp-calypso/pull/31524#pullrequestreview-223300577) in which I approved this PR, I agree to disagree on this point.

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +52,88 @@ class SiteTypeForm extends Component {
+				<p className="site-type__other-label">{ translate( 'Or type your own' ) }</p>
+
+				<div className="site-type__other-form">
+					<FormTextInput
```
For better accessibility, I should be able to press "enter" and submit this.

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +52,88 @@ class SiteTypeForm extends Component {
+				<p className="site-type__other-label">{ translate( 'Or type your own' ) }</p>
+
+				<div className="site-type__other-form">
+					<FormTextInput
```
And another question: should we be limiting the number of characters with a `maxlength` attribute here? Would be nice for the user so they know they're limited (although it is quite an edge case).

AnnaMag replied:
I do not think it is necessary.

tyxla replied:
Well, why not? We're limiting in the backend, so it only makes sense to limit in the frontend too, for consistency at least.

Also, I still think we could harness `FormTextInputWithAction` and remove some code.

AnnaMag replied:
I do no think it is necessary atm since the design speaks for itself. If upon analyzing the results we see that the input is too lengthy, I'd add it.

tyxla replied:
I don't see how a `maxlength` could hurt, but if you have so hard feelings to not add it, that's fine.

What are your thoughts about using `FormTextInputWithAction`?

AnnaMag commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +52,88 @@ class SiteTypeForm extends Component {
+
+					<Button
+						className="site-type__other-submit"
+						disabled={ false }
```
Actually, the goal is to have the opposite happen—the button can be used without input and with no text specified, save "unspecified" value. I adapted the code accordingly.

AnnaMag commented:
> For better accessibility, I should be able to press "enter" and submit this.

Great point, I will ship it in the follow up next week. 

I addressed all feedback, and applied wherever relevant. The improvements to CSS were not included as the code is to be removed shortly.  Merging as we should start collecting the data asap.

tyxla commented:
I'd love to give this another spin before merging, if you don't mind.

AnnaMag commented:
@tyxla I just pushed another correction. Please feel free. If there is anything you feel like needs a change, please push a commit directly. The goal is to collect the data over the weekend, even if not all code is not perfect (it is to be removed).
I will hold off with the merge till the evening.

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -30,10 +34,18 @@ class SiteTypeForm extends Component {
 	constructor( props ) {
 		super( props );
 		this.state = {
+			hasOtherReasonFocus: false,
```
This is never used.

AnnaMag replied:
removed, thank you

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +56,80 @@ class SiteTypeForm extends Component {
-				<strong className="site-type__option-label">{ siteTypeProperties.label }</strong>
-				<span className="site-type__option-description">{ siteTypeProperties.description }</span>
+	handleSubmitOther = () => {
+		if ( ! this.state.otherValue || this.state.hasOtherReasonFocus ) {
```
`this.state.hasOtherReasonFocus` will never be true. Seems like it can be removed.

AnnaMag replied:
This is older commit, so had been removed already before.

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +56,80 @@ class SiteTypeForm extends Component {
+		this.handleSubmit( 'other—' + this.state.otherValue );
+	};
+
+	renderBasicCard = () => {
```
Nit: No need to be an arrow function.

tyxla commented on client/signup/steps/site-type/form.jsx:
```diff
@@ -44,29 +56,80 @@ class SiteTypeForm extends Component {
+		);
+	};
+
+	renderOtherInfo = () => {
```
Nit: No need to be an arrow function.
